### ISSUE TITLE
(8) URGENT: Reconsider optimization association with arrays

### ISSUE THREAD

**Issue by thomasrolinger** [thomasrolinger/chapel#12]
This is a fundamental decision that needs to be made before going forward with anything else.

I've been avoiding associating the optimization (ieObj, replArrays) with particular array instances. This is because the real use-cases I had did not have stand-alone `B` arrays, but rather fields of a record. This makes it difficult to consider multiple instances of `B`. The reason why we do this is because `B` is often a completely different size versus `A`, so the block distribution is out of sync.

However, I ran some experiments using "stand alone" indirection arrays and the performance difference is very little, if nothing at all. This is because of Chapel's remote cache, which does a good job on prefetching the sequential accesses to the indirection array.

With that said, there is little reason to cling to this idea of separating the optimization from the arrays. If anything, it is causing more complications than it is solving.

Below is a rough plan of what we will need to do:
* Move `replArrays` and `ieObj` down into the array itself rather than having them as globals. `A` will have `replArrays` and an `ieObj` instance that will keep track of changes to its domain. `B` will have an `ieObj` instance as well to keep track of changes to it and its domain. We will also give the domains an `ieObj` so that changes to that can be more easily tracked.
* Require that both `A` and `B` are defined outside of the `forall`.
* Relax the requirement that the loop contains `A[B[i]]` must iterate over `B.domain`. In our real codes, we do not iterate directly/entirely over `B.domain` but rather a subset. This could be done via slicing (`B.domain[1..n]`) or creating a subdomain of `B.domain`. So we should detect at least one of these situations.
* The code we generate will now look like
```chpl
if doInspector(A) || doInspector(B) || doInspector(A.domain) || doInspector(B.domain) {
   inspector loop
   sortIndices(A);
   inspectorOff(A);
   inspectorOff(B);
}
executorPreamble
executor loop
```
* For tracking changes to the arrays and their domains, we will want to try to use the same approach we have. Alternatively, we could flip the flag on any write to any array/domain. But the problem with that is some writes to `A` are completely valid and do not need to restart the inspector. So we will need to start from the forall and branch outwards to look for valid writes. The problem with that is we still need to address the multiple callsites/same call sites with different arrays issue. We'll need some extra analysis to find all possible arrays that would be passed into the forall and track those.

**Comment by thomasrolinger** [comment]
Another potential approach is to keep things the way they are, but do more aggressive/extension function cloning. Consider the following case:

```chpl
proc foo(A, B)
{
   forall i in B.domain {
      A[B[i]];
   }
}

proc bar(X, Y, W, Z)
{
   foo(X, Y);
   foo(W, Z);
}

proc beef(H, I, J, K, M, N, O, P)
{
   bar(H, I, J, K);
   bar(M, N, O, P);
}

beef(T, D, G, C, U, V, S, Q);
```

What we can do is work backwards from the `forall` in `foo` to construct the possible call chains:
```
beef -> bar (H, I..) -> foo (X,Y)
beef -> bar (H, I..) -> foo (W, Z)
beef -> bar (M, N..) -> foo (X, Y)
beef -> bar (M, N..) -> foo (W, Z)
```
What we need to do is create clones for each "duplicate" call site we see. In this case, that means cloning `bar` into `bar1` and `bar2` and `foo` into `foo1`, `foo2`, `foo3`, and `foo4`. The exact mechanism we use to do this is not immediately clear, but it seems do-able.

**Comment by thomasrolinger** [comment]
This has been addressed: https://github.com/thomasrolinger/chapel/commit/4bc6d18c0b1eb54a932d4c5028f7c112197ca693

Code generation works but we still need to go back and fix up some of the static analysis.

* We added an `inspectorFlag` field to the `BlockArr` and `BlockDom` classes. Each locale gets one of these instances, so each locale has its own flag. This is OK, as when we call `inspectorOn()` and `inspectorOff()` we set the flag on all locales.
* For now, the flag is initialized to true and we check for whether `A`'s flag is true. Even though we do not track modifications to `A`, this allows us to handle multiple call sites with different `A` arrays. If we didn't check for `A`'s flag and only `B`'s and its domain, then after the first call site, we would not do the inspector for the second one, which leads to an issue.
* We add a `ReplicatedElems` record as a field to the `BlockArr` and `BlockDom` classes. It functions in a similar way to the `inspectorFlag`. Each locale has its own record, which is what we want.

From there, we made changes to the compiler portion of the optimization:
* Code generation is done almost entirely in normalization, with the exception of calling `executorPreamble` since it relies on knowing whether we are dealing with records or not.
* We can now use `forall` intents on the inspector and executor loops.

What we still need to do is go back and update/revise is some of the static analysis:
* Static analysis will need to be updated to not allow for `B` to be defined in the loop
* Static analysis will need to be updated to account for the loop potentially iterating over a subdomain of `B.domain`. We'll need mechanisms to detect and track it appropriately.
* Still need to clone call sites so that if the function is not called from a valid place, it does not call into the optimized version. We can either make a clone and make it the optimized version, or make a clone and make it the original version.
